### PR TITLE
feature(example): Now builds a Docker image thanks to a local Dockerfile

### PR DESCRIPTION
**Pull request by ash-sxn**
* added Dockerfile to customize the image to hide the complexity (idea from @jmMeessen )
* addedJcasc file to configure Jenkins( [#1  really helped for guidance, Thank you @gounthar )
* added plugins.txt list for jenkins-plugin-cli
* and a 02_docker-compose.yaml with build instead of image

### CODE REVIEW

**Comment by ash-sxn** [comment]
pinging
@jmMeessen @berviantoleo @gounthar

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM jenkins/jenkins:2.401.1-lts
+USER root
+RUN echo "2.0" > /usr/share/jenkins/ref/jenkins.install.UpgradeWizard.state
+COPY plugins.txt /usr/share/jenkins/ref/plugins.txt
+RUN jenkins-plugin-cli --plugin-file /usr/share/jenkins/ref/plugins.txt
+COPY jenkins.yaml /var/jenkins_home/casc_configs/jenkins.yaml
+ENV CASC_JENKINS_CONFIG /var/jenkins_home/casc_configs/jenkins.yaml
```
Comment (berviantoleo) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/Dockerfile:7]: Please set back the user to "jenkins" as mentioned [here](https://github.com/jenkinsci/docker#installing-more-tools)

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/jenkins.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+jenkins:
```
Comment (berviantoleo) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/jenkins.yaml:1]: I think we can disable the built-in agent through this property.

```yaml
jenkins:
    numExecutors: 0
```

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+gradle
+workflow-aggregator
+github-branch-source
+pipeline-stage-view
+git
+matrix-auth
+pam-auth
+pipeline-github-lib
+ssh-agent
+ldap
```
Comment (berviantoleo) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt:19]: Please remove unused plugins.

**Comment by jmMeessen** [comment]
The system proposed by the PR ( 02_custom_docker_file_connecting_agent_and_controller) works 👍 

Important improvements on top of those of Bervianto:
As the system's complexity will increase, it is important that the directory structure help to better understand what is happening.

1. I therefore recommend that you store all the material required to build the custom docker image in a special sub-directory (ex `./images/controler/` or `./dockerfiles/` or `./dockerfiles/controler/`)
2. store the security related stuff (ex the secret key) in its dedicated directory (I use `./secrets/`)
3. subject to discussion: importing the JCasC configuration in the image (dockerfile) makes it very static (which can make sense). The alternative way would be to map a local directory (containing the config files) into the adequate place in the Jenkins filesystem. It will make it very easy to adapt the configuration.
4. we need to discuss the rational and implication for mounting a named volume. The advantage is that (manual) modifications to the configuration are persistant. But is this something that we want to have. To restart the system, it requires to remove the named volume (a `docker compose down` is not sufficient).
5. I have a feeling that adding a small note (aka readme.me) in each experiment would help people like me a lot. You could start by saying what does the demo do (or is supposed to do), how to start it, how to tear it down.

/- Jmm

**Comment by ash-sxn** [comment]
> 1. I therefore recommend that you store all the material required to build the custom docker image in a special sub-directory (ex `./images/controler/` or `./dockerfiles/` or `./dockerfiles/controler/`)
> 2. store the security related stuff (ex the secret key) in its dedicated directory (I use `./secrets/`)
> 3. subject to discussion: importing the JCasC configuration in the image (dockerfile) makes it very static (which can make sense). The alternative way would be to map a local directory (containing the config files) into the adequate place in the Jenkins filesystem. It will make it very easy to adapt the configuration.
> 4. we need to discuss the rational and implication for mounting a named volume. The advantage is that (manual) modifications to the configuration are persistant. But is this something that we want to have. To restart the system, it requires to remove the named volume (a `docker compose down` is not sufficient).
> 5. I have a feeling that adding a small note (aka readme.me) in each experiment would help people like me a lot. You could start by saying what does the demo do (or is supposed to do), how to start it, how to tear it down.

Thanks @jmMeessen 
I'll create separate PR for 1st 2nd and 5th point (after asking some doughts on the Gitter channel )
I do agree with you about 3rd point, that it makes it very static
on 4th point, will we use bind mount then? 
What do you guys think @gounthar @berviantoleo

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+configuration-as-code
```
[the remark this answers is not in the document]
Comment (ash-sxn) [reply experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt]: I've made the changes Please take a look

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt (hunk 3 above)
Comment (ash-sxn) [reply experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt]: I've made the changes Please take a look

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/Dockerfile (hunk 1 above)
Comment (ash-sxn) [reply experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/Dockerfile:7]: I've made the changes Please take a look

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt (hunk 3 above)
Comment (berviantoleo) [reply experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/plugins.txt]: Good job.

**Comment by berviantoleo** [comment]
Please proceed with the 1st, 2nd, and 5th points.

Basically, 5th point is what I'm looking for in #13.

For 4th point, I'm okay with any approach.

**Comment by ash-sxn** [comment]
I'll create separate ReadMe files for each example

**Comment by ash-sxn** [comment]
I've created a README.md file for 02 example here since this PR is not merged right now, but will create separate PR for other examples

**Comment by ash-sxn** [comment]
Instead of creating  separate PR of 1st and 2nd points of @jmMeessen suggestion, I've added it here in this PR

**Review by berviantoleo — APPROVED** [review]

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+## Example 2
+### Custom Docker File connecting agent and controller 
+This example builds a custom docker image to skip the upgrade wizard, install default plugins, and configure Jenkins with JCasC.
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:3]:
```suggestion
This example builds a custom docker image to skip the upgrade wizard, installs default plugins, and configures Jenkins with JCasC.
```

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+### Files 
+* Dockerfile is used to create a custom image from the `jenkins/jenkins` image and copy plugins.txt and jenkins.yaml into the container for plugin installation and JCasC configuration.
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:5]:
```suggestion
* The Dockerfile is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
```

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+* jenkins_agent_ed is an ed25519 private ssh key used in the setup.
+* jenkins.yaml is a JCasC configuration file that can be edited to customize the build accordingly.
+* plugins.txt is a list of plugins that are installed with the help of jenkins-plugin-cli.
+* The docker-compose.yaml file defines 2 containers, the Jenkins controller and agent, connects them to a network, gets the private key in the controller and public key in the agent, defines a named volume, and also builds the image from the Dockerfile.
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:9]:
```suggestion
* `jenkins_agent_ed` is an `ed25519` private ssh key used in the setup.
*` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
```

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+### Steps to run the example 
+* Make sure you are in the `02_custom_docker_file_connecting_agent_and_controller` directory by using the command `pwd`. 
+* Then use the command `docker compose up -d` to run the example.
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:13]:
```suggestion
* Then use the command `docker compose up -d --build` to run the example.
```

**Review by gounthar — CHANGES_REQUESTED** [review]
I think we should tell what the credentials are.

**Comment by ash-sxn** [comment]
> I think we should tell what the credentials are.

Login credentials for instance, right?
that is admin, admin for this example,

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+### Files 
+* The Dockerfile is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
+* `jenkins_agent_ed` is an `ed25519` private ssh key used in the setup.
+*` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
+* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
+* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
+
+### Steps to run the example 
+* Make sure you are in the `02_custom_docker_file_connecting_agent_and_controller` directory by using the command `pwd`. 
+* Then use the command `docker compose up -d --build` to run the example.
```
Comment (berviantoleo) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:13]: Please add the steps to login in here and mention the credentials.

**Comment by gounthar** [comment]
> > I think we should tell what the credentials are.
> 
> Login credentials for instance, right? that is admin, admin for this example,

Yes, but not everyone will guess it. 🤷 
Better safe than sorry. ;-)

**Comment by jmMeessen** [comment]
We are reaching a good state, I feel. several things can/still need to be added to the readme (for example, how to cleanly destroy the composed environment). I proposed that, once the requested changes have been implemented, we merge this PR. It will clarify the work in progress.

**Review by jmMeessen — APPROVED** [review]
We have a first acceptable iteration (once the other requested changes have been implemented).

**Comment by ash-sxn** [comment]
Sorry for the delay @gounthar 
I've added the credentials

**Review by berviantoleo — APPROVED** [review]

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+* You can login using the username `admin` and password `admin`
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:14]:
```suggestion
* You can login using the username `admin` and password `admin````

Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:14]:
```suggestion
* You can login using the username `admin` and password `admin`.
```

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+* The Dockerfile is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
+* `jenkins_agent_ed` is an `ed25519` private ssh key used in the setup.
+*` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
+* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
+* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
+
+### Steps to run the example 
+* Make sure you are in the `02_custom_docker_file_connecting_agent_and_controller` directory by using the command `pwd`. 
+* Then use the command `docker compose up -d --build` to run the example.
+* You can login using the username `admin` and password `admin`.
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:14]:
```suggestion
* You can log in using the username `admin` and password `admin`.
```

File: experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+* The Dockerfile is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
+* `jenkins_agent_ed` is an `ed25519` private ssh key used in the setup.
+*` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
+* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
+* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
+
+### Steps to run the example 
+* Make sure you are in the `02_custom_docker_file_connecting_agent_and_controller` directory by using the command `pwd`. 
+* Then use the command `docker compose up -d --build` to run the example.
+* You can log in using the username `admin` and password `admin`.
```
Comment (gounthar) [inline experimental_docker_compose_files/02_custom_docker_file_connecting_agent_and_controller/README.md:14]:
```suggestion
* You can log in using the `admin` username and the `admin` password.
```

**Review by gounthar — APPROVED** [review]
Fine with me, thanks.

**Pull request merged**